src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1672,39 +1672,61 @@ public boolean isCastable(Type t, Type s, Warner warn) {
-                     * an interface is disjoint from a class if the class is disjoint form the interface
-                     */
-                    return areDisjoint(ss, ts);
-                }
-                // a final class that is not subtype of ss is disjoint
-                if (!ts.isInterface() && ts.isFinal()) {
-                    return true;
-                }
-                // if at least one is sealed
-                if (ts.isSealed() || ss.isSealed()) {
-                    // permitted subtypes have to be disjoint with the other symbol
-                    ClassSymbol sealedOne = ts.isSealed() ? ts : ss;
-                    ClassSymbol other = sealedOne == ts ? ss : ts;
-                    return sealedOne.getPermittedSubclasses().stream().allMatch(type -> areDisjoint((ClassSymbol)type.tsym, other));
+
+                if (ts.isInterface() && !ss.isInterface() || // case I: one is a class and the other one is an interface
+                    !ts.isInterface() && ss.isInterface()) {
+                    ClassSymbol isym = ts.isInterface() ? ts : ss; // isym is the interface and csym the class
+                    ClassSymbol csym = isym == ts ? ss : ts;
+                    if (!isSubtype(erasure(csym.type), erasure(isym.type))) {
+                        if (csym.isFinal()) {
+                            return true;
+                        } else if (csym.isSealed()) {
+                            return areDisjoint(isym, csym.getPermittedSubclasses());
+                        } else if (isClassFreelyExtensible(csym) && isym.isSealed()) {
+                            return areDisjoint(csym, isym.getPermittedSubclasses());
+                        }
+                    }
+                } else if (!ts.isInterface() &&              // case II: both are classes
+                        !ss.isInterface()) {
+                    return !isSubtype(erasure(ss.type), erasure(ts.type)) && !isSubtype(erasure(ts.type), erasure(ss.type));
+                } else {                                     // case III: both are interfaces
+                    if (!isSubtype(erasure(ts.type), erasure(ss.type)) && !isSubtype(erasure(ss.type), erasure(ts.type))) {
+                        if (ts.isSealed()) {
+                            return areDisjoint(ss, ts.getPermittedSubclasses());
+                        } else if (ss.isSealed()) {
+                            return areDisjoint(ts, ss.getPermittedSubclasses());
+                        }
+                    }
                 }
+                // at this point we haven't been able to prove that the classes or interfaces are disjoint so we bail out
                 return false;
             }
+
+            private boolean isClassFreelyExtensible(ClassSymbol csym) {
+                boolean anySuperIsSealed = (csym.getSuperclass().tsym != null && csym.getSuperclass().tsym.isSealed()) || csym.getInterfaces().stream().anyMatch(i -> i.tsym.isSealed());
```

Comment:
I am not sure about these checks. Should it be simply `isClassFreelyExtensible == !csym.isSealed() && !csym.isFinal()`? Or is the a particular case this is solving.

(Besides this code being more complex, I suspect "non-sealed" is not restored when reading from classfiles - is that correct? Will this code do something sensible when `csym` is read from a classfile, and has a sealed supertype?)

Reply:
yes I think you are right, will do